dace/sdfg/analysis/writeset_underapproximation.py

```diff
@@ -689,21 +677,34 @@ def _merge_subsets(subset_a: subsets.Subset, subset_b: subsets.Subset) -> subset
         return subset_b
 
 
+class UnderapproximateWritesDictT(TypedDict):
+    approximation: Dict[graph.Edge, Memlet]
+    loop_approximation: Dict[SDFGState, Dict[str, Memlet]]
+    loops: Dict[SDFGState, Tuple[SDFGState, SDFGState, List[SDFGState], str, subsets.Range]]
+
+
+@transformation.experimental_cfg_block_compatible
 class UnderapproximateWrites(ppl.Pass):
 
+    # Dictionary mapping each edge to a copy of the memlet of that edge with its write set underapproximated.
+    approximation_dict: Dict[graph.Edge, Memlet] = {}
+    # Dictionary that maps loop headers to "border memlets" that are written to in the corresponding loop.
+    loop_write_dict: Dict[SDFGState, Dict[str, Memlet]] = {}
+    # Dictionary containing information about the for loops in the SDFG.
+    loop_dict: Dict[SDFGState, Tuple[SDFGState, SDFGState, List[SDFGState], str, subsets.Range]] = {}
+    # Dictionary mapping each nested SDFG to the iteration variables surrounding it.
+    iteration_variables: Dict[SDFG, Set[str]] = {}
+    # Mapping of state to the iteration variables surrounding them, including the ones from surrounding SDFGs.
+    ranges_per_state: Dict[SDFGState, Dict[str, subsets.Range]] = defaultdict(lambda: {})
```

Comment:
the default initialization here is wrong. You should either use dataclasses.field if you want that, or just not initialize them here and add an `__init__` that calls `super().__init__()` and then initializes those.